cli/tests/test_git_push.rs

```diff
@@ -2244,6 +2244,88 @@ fn test_git_push_rejected_by_remote() {
+    [revsets]
+    'git-push' = "remote_bookmarks(remote=remote)..@ ~untracked_remote_bookmarks(glob:'local/*')"
+    "#,
+    );
+    work_dir
+        .run_jj(["new", "bookmark2", "-m", "commit to be pushed"])
+        .success();
+    work_dir.run_jj(["new", "-m", "wip: stuff"]).success();
+    work_dir
+        .run_jj(["bookmark", "set", "local/stuff", "-r@"])
+        .success();
+    work_dir
+        .run_jj(["new", "-m", "commit which should pushed"])
+        .success();
+    //
+    let output = work_dir.run_jj(["log"]);
+    insta::assert_snapshot!(output, @r"
+    @  kpqxywon test.user@example.com 2001-02-03 08:05:17 048a6554
+    │  (empty) commit which should pushed
+    ○  yostqsxw test.user@example.com 2001-02-03 08:05:15 local/stuff 0712d559
+    │  (empty) wip: stuff
+    ○  vruxwmqv test.user@example.com 2001-02-03 08:05:14 a615282a
+    │  (empty) commit to be pushed
+    ○  zsuskuln test.user@example.com 2001-02-03 08:05:10 bookmark2 38a20473
+    │  (empty) description 2
+    │ ○  qpvuntsm test.user@example.com 2001-02-03 08:05:08 bookmark1 9b2e76de
+    ├─╯  (empty) description 1
+    ◆  zzzzzzzz root() 00000000
+    [EOF]
+    ");
+    // We should try to push everything except the commit on the "local/" bookmark.
+    let output = work_dir.run_jj(["git", "push"]);
+    insta::assert_snapshot!(output, @r"
+    ------- stderr -------
+    Warning: Refusing to create new remote bookmark local/stuff@origin
+    Hint: Use --allow-new to push new bookmark. Use --remote to specify the remote to push to.
+    Nothing changed.
+    [EOF]
+    ");
+}
+
+// This tests that `exact: <remote>` isn't changed to a substring pattern.
+#[test]
+fn test_git_push_unambigous_remote() {
+    let test_env = TestEnvironment::default();
+    set_up(&test_env);
+    let work_dir = test_env.work_dir("local");
+    work_dir.run_jj(["new", "bookmark1"]).success();
+    let add_new_git_remote = |remote: &str| {
+        test_env.run_jj_in(".", ["git", "init", remote]).success();
+        let origin_dir = test_env.work_dir(remote);
+        let remote_dir = git_repo_dir_for_jj_repo(&origin_dir);
+        work_dir
+            .run_jj(["git", "remote", "add", remote, remote_dir.to_str().unwrap()])
+            .success();
+    };
+    // Add some new remotes which could get matched on substrings.
+    add_new_git_remote("upstream");
+    add_new_git_remote("stream");
+    work_dir.run_jj(["new", "bookmark1"]).success();
+    work_dir.run_jj(["desc", "-m", "Foo"]).success();
+    work_dir.run_jj(["bookmark", "move", "bookmark1"]).success();
+    // This will push to two remotes if we use substring matching.
+    let output = work_dir.run_jj(&["git", "push", "--remote", "stream"]);
+    insta::assert_snapshot!(output, @r"
+    ------- stderr -------
+    Warning: Refusing to create new remote bookmark bookmark1@stream
+    Hint: Use --allow-new to push new bookmark. Use --remote to specify the remote to push to.
+    Nothing changed.
+    [EOF]
```

Comment:
This wouldn't fail with substring match. We'll need to set up two remotes with the same bookmark pointing to different commits, and push it to the other remote, something like:
```
foo@stream foo
|
foo@upstream
```

Reply:
I tried to rewrite the test around this but it currently doesn't look right yet.